troubleshooting/kubernetes/Kubernetes_Kube_apiserver_failing.md

```diff
@@ -0,0 +1,131 @@
+        Expected output if encryption is enabled but has not been restored on a single master node:
+
+        ```bash
+        ncn-m001:~ # /usr/share/doc/csm/scripts/operations/kubernetes/encryption.sh --status
+        k8s encryption status
+        changed: 2024-07-06 20:07:35+0000
+        ncn-m001: identity
+        ncn-m002: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        ncn-m003: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        current: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        goal: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        etcd: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        interim state detected, ensure all control plane nodes are in sync
+        ```
+
+        Expected output if encryption is enabled on all master nodes:
+
+        ```bash
+        ncn-m001:~ # /usr/share/doc/csm/scripts/operations/kubernetes/encryption.sh --status
+        k8s encryption status
+        changed: 2024-07-06 20:07:35+0000
+        ncn-m001: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        ncn-m002: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        ncn-m003: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        current: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        goal: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        etcd: aescbc-625e61a4ebe4d3ddf8b5eec3b546663945b837d53ca966d72e49b42cdae4e656 identity
+        ```
+
+    If Kubernetes encryption is set up on the system and is enabled on all master nodes, then there is nothing more to do. Follow the next step if Kubernetes encryption is not correctly set up on a master node.
+
+1. (`ncn-m001`) Set up Kubernetes encryption on the master node that does not have it enabled.
+
+    1. Copy `/etc/cray/kubernetes/encryption` files to the master node without Kubernetes encryption enabled from another master node. Adjust the node names below based on which master node contains the correct files and which node is the target node.
+
+        ```bash
+        scp /etc/cray/kubernetes/encryption/* ncn-m001:/etc/cray/kubernetes/encryption/
+        ```
+
+    1. Restart `kube-apiserver` on the node where Kubernetes encryption is being enabled. This should be the same node that the `kube-apiserver` is failing on and why this troubleshooting document is being followed.
+
+        Set the `failing_api_serv_node` to the node where the `kube-apiserver` is failing.
+
+        ```bash
+        failing_api_serv_node=ncn-m001
+        ```
+
+        Delete the failing `kube-apiserver` pod.
+
+        ```bash
+        kubectl delete pod -n kube-system kube-apiserver-${failing_api_serv_node}
+        ```
```

Comment:
Unless something has changed, I don't think that will work.  The `kube-apiserver` pod is a static pod, meaning it is started directly by the kubelet via the manifest in `/etc/kubernetes/manifests/kube-apiserver.yaml`, so what is being shown with kubectl is just a shadow pod.  Deleting the shadow pod doesn't actually restart the `kube-apiserver`.  On the master node where you want to restart `kube-apiserver`, you have to do something like:
```
cd /etc/kubernetes/manifests
mv kube-apiserver.yaml ..
```
The kubelet will notice that the manifest is gone, and will stop the pod.  You can verify that with `kubectl`, or using `crictl ps | grep kube-apiserver` to see that it is no longer running.  Once it is stopped,
```
cd /etc/kubernetes/manifests
mv ../kube-apiserver.yaml .
```
and the kubelet will notice the manifest and restart the kubelet.

Reply:
Here's an example showing that if you just delete the shadow pod, the underlying container does not get restarted:
```
k8s-master-sigma7-1:~ # kubectl get pods -A | grep kube-apiserver
kube-system      kube-apiserver-k8s-master-sigma7-1                                1/1     Running                 2 (83s ago)        14d
kube-system      kube-apiserver-k8s-master-sigma7-2                                1/1     Running                 0                  14d
kube-system      kube-apiserver-k8s-master-sigma7-3                                1/1     Running                 0                  14d
k8s-master-sigma7-1:~ # crictl ps | grep kube-apiserver
53e6526434f0e       4f1c5007cffae       3 minutes ago       Running             kube-apiserver            2                   8665492883c01       kube-apiserver-k8s-master-sigma7-1
k8s-master-sigma7-1:~ # kubectl delete pod -n kube-system      kube-apiserver-k8s-master-sigma7-1
pod "kube-apiserver-k8s-master-sigma7-1" deleted
k8s-master-sigma7-1:~ # kubectl get pods -A | grep kube-apiserver
kube-system      kube-apiserver-k8s-master-sigma7-1                                0/1     Pending                 0                  3m50s
kube-system      kube-apiserver-k8s-master-sigma7-2                                1/1     Running                 0                  14d
kube-system      kube-apiserver-k8s-master-sigma7-3                                1/1     Running                 0                  14d
k8s-master-sigma7-1:~ # kubectl get pods -A | grep kube-apiserver
kube-system      kube-apiserver-k8s-master-sigma7-1                                1/1     Running                 2 (4m37s ago)      4m1s
kube-system      kube-apiserver-k8s-master-sigma7-2                                1/1     Running                 0                  14d
kube-system      kube-apiserver-k8s-master-sigma7-3                                1/1     Running                 0                  14d
k8s-master-sigma7-1:~ # crictl ps | grep kube-apiserver
53e6526434f0e       4f1c5007cffae       4 minutes ago       Running             kube-apiserver            2                   8665492883c01       kube-apiserver-k8s-master-sigma7-1
k8s-master-sigma7-1:~ # 
```
Note that the underlying container has not changed (same ID, same run time), all that happened is that the shadow pod got re-created.

Reply:
Thanks for your comments. I have fixed the command and used the 'crictl' option. I have tested restoring encryption multiple times with this procedure and it has worked.